### PR TITLE
Fixed loading of scripts in pjax containers

### LINKED ISSUE #23: Problem with page-specific JS assetbundles
I have been implementing PJAX using the yii2-pjax plugin quite successfully. I'm using it to only load the content on my page and keep the header and sidebar static, saving quite a few seconds on not having to reload the assets added to the header and the sidebar.

However, on some of my pages I'm using certain asset bundles to load in javascript only applied to those pages. When loading in the content with pjax these asset bundles are not being registered as I expected they would be. I'm wondering if there is any way to do what I want to do or if the only solution is to add a global javascript file that includes all functions? I would rather not do this as this might create quite a big file and a lot of unused code on certain pages.

This is how I've defined my content:

        <?php Pjax::begin() ?>
             AdminAsset::register($this);
             // Content here
        <?php Pjax::end() ?>

I've looked for the majority of the day and can't find a clean solution besides having one big application.js file or using inline javascript, both of which I would like to avoid.

### PR DESCRIPTION
**Pull request by SilverFire**
Closes yiisoft/yii2#3680
Closes yiisoft/yii2#4814
Closes yiisoft/yii2#8540
Closes yiisoft/yii2#8916
Closes yiisoft/yii2#8702
Closes #23
Closes #22

This patch was proposed by @nkovacs. I've tested it for 2 week in my project and found no problems with it.

May be merged if nobody complains.

### CODE REVIEW

File: jquery.pjax.js [hunk 1]
```diff
@@ -785,26 +785,53 @@ function extractContainer(data, xhr, options) {
...
-    if (matchedScripts.length) return
 
-    var script = document.createElement('script')
-    var type = $(this).attr('type')
-    if (type) script.type = type
-    script.src = $(this).attr('src')
-    document.head.appendChild(script)
-  })
+    if (matchedScripts.length) {
+      next()
```
Comment (samdark) [inline jquery.pjax.js]: It's better to use `;`

**Comment by nkovacs** [comment]
Instead of using onload/onerror, you could also use an ajax call and jquery's globalEval, like jquery._evalUrl, except with an async ajax call and the callbacks.
Btw, there's a proposed patch to fix it, but the original pull request is broken, there's a better pseudo-implementation in the comments: https://github.com/jquery/jquery/pull/2126

**Comment by SilverFire** [comment]
@nkovacs did you test it?

**Comment by nkovacs** [comment]
Not with this, no. I monkey patched $.domManip, and it works.
This is how globalEval works:
```
if ( code.indexOf("use strict") === 1 ) {
    script = document.createElement("script");
    script.text = code;
    document.head.appendChild( script ).parentNode.removeChild( script );
} else {
    indirect( code );
}
```
I'm not sure document.head.appendChild is guaranteed to be synchronous, maybe not, but in that case it's strange that jquery uses it. The else branch uses eval, so that's synchronous, and in my case the code doesn't have 'use strict' (yii's activeform and validation js), so it works for me.

The onload thing might not work on old IE, but since pjax only works on browsers supporting history api, that's not an issue.

**Comment by SilverFire** [comment]
But your implementation in this PR works pretty good, isn't it? Do you think it must be changed?

**Comment by nkovacs** [comment]
I don't know which one is better. Onload works in practice on the modern browser I tried, not sure how reliable it is. Eval should be good, but it's unsafe, and not allowed in many places.

**Comment by SilverFire** [comment]
Okay, I'm preferring implementation in this PR as it is tested

**Comment by mikehaertl** [comment]
@SilverFire What about the issue in #6655? Shouldn't it also try to address this problem: https://github.com/yiisoft/yii2/issues/6655#issuecomment-149196252?

**Comment by SilverFire** [comment]
@mikehaertl yes, your idea makes sense. Will create a PR in yii2 repo later

**Comment by mikehaertl** [comment]
@SilverFire So just to be clear: you want to unify the script filter mechanism, right? Because to me that's the root of many problems: Pjax and Yii both use their own way to make sure, that a script is not loaded twice. But these two ways are not compatible.

So maybe we should refactor Yii's script filter and use the same mechanism like here. Ideally we should even reuse the same code...

**Pull request merged**